Firmware/LowLevel/src/main.cpp

```diff
@@ -547,13 +548,56 @@ void onUIPacketReceived(const uint8_t *buffer, size_t size) {
     }
 }
 
-void sendConfigMessage(uint8_t pkt_type) {
-    struct ll_high_level_config ll_config;
-    ll_config.type = pkt_type;
-    ll_config.config_bitmask = config_bitmask;
-    ll_config.volume = 80;            // FIXME: Adapt once nv_config or improve-sound got merged
-    strncpy(ll_config.language, "en", 2); // FIXME: Adapt once nv_config or improve-sound got merged
-    sendMessage(&ll_config, sizeof(struct ll_high_level_config));
+void sendConfigMessage(const uint8_t pkt_type) {
+    size_t msg_size = sizeof(struct ll_high_level_config) + 3;  // + 1 type + 2 crc
+    uint8_t *msg = (uint8_t *)malloc(msg_size);
+    if (msg == NULL)
+        return;
+    *msg = pkt_type;
+    memcpy(msg + 1, &llhl_config, sizeof(struct ll_high_level_config));  // Copy our live config into the message, behind type
+    sendMessage(msg, msg_size);
+    free(msg);
+}
+
+void applyConfig(const uint8_t *buffer, const size_t size) {
+    // This is a flexible length packet where the size may vary when ll_high_level_config struct got enhanced only on one side.
+    // If payload size is larger than our struct size, ensure that we only copy those we know of = our struct size.
+    // If payload size is smaller than our struct size, copy only the payload we got, but ensure that the unsent member(s) have reasonable defaults.
+    size_t payload_size = min(sizeof(ll_high_level_config), size - 2);  // exclude crc
+
+    // Use a temporary config for easier sanity adaption and copy our live config, which has at least reasonable defaults.
+    // The live config copy ensures that we've reasonable values for the case that HL config struct is older (smaller) than ours.
+    auto hl_config = llhl_config;
+
+    // Copy payload to temporary config
+    memcpy(&hl_config, buffer, payload_size);
+
+    // Takeover of HL leading config values
```

Comment:
Mind explaining what all this logic for "HL leading" is about? I thought that HL would always send the full config and we take over everything (expect on version mismatch, but I think we can ignore that for a second). You rather want to be able to send a delta config, which keeps "undefined" fields as they are? Is that really necessary and worth the complexity? I think you mentioned volume which can be set on the mower, but is there anything else that would be controlled from LL?

Reply:
Yes sure, will try, might be that it sounds somehow confusing :smile: 

At the moment we need to maintain/align a couple of LL configs settings also in HL. In my point of view this is error prone and redundant, as well as HL should not care much about LL config values/defaults.

I know, this sounds contrary to what I plan to implement. But we do have two use case scenarios.
\>90?% will go with the default LL settings whereas some others like to use adapted ones.

My target is that HL only inform LL about the adapted ones, whereas LL inform about the used (default) ones.

Yesterday I build parts of that and I need to say that it also look somehow ugly as well as confusing to me :disappointed: ... later in bed I recognized that I probably only need to change the config- triggering side from LL to HL which is much more logic as well as it save some stupid looking assignments.

Now that the protocol is done in general and the communication work, I'll try a little bit with one value and look how it could be made more straightforward. Will give you an code example when I got it together in a more clear way.

Reply:
Most of this is about the defaults though? Apart from volume, is there any other setting that is changeable from the mower side?

Reply:
Yes. In my vision also all other values like charge, battery, timeouts, ... if not separately defined in HL (mower_config).

I'm almost done with my implementation. Let's simply wait one or two more days if I'm on the wrong track. If not, code says more than 1000 words :laughing: 

BTW: Think I know how to get rid of `IGNORE_CHARGING_CURRENT`. How's about giving it a bit in https://github.com/Apehaenger/OpenMower/blob/37f2ac6d2f30e52d7e6433d08c7101fa9e7e4b48/Firmware/LowLevel/src/datatypes.h#L177 ?

Reply:
> Yes. In my vision also all other values like charge, battery, timeouts, ... if not separately defined in HL (mower_config).

But those wouldn't be changed "live"/at runtime by pressing some button on the mower, right?
I think you're saying that you don't want to maintain the defaults in two different places, which I think is a good point. I wanted to make up my mind about this, and therefore wanted to know how "dynamic" the config is on LL side.

> BTW: Think I know how to get rid of `IGNORE_CHARGING_CURRENT`. How's about giving it a bit in https://github.com/Apehaenger/OpenMower/blob/37f2ac6d2f30e52d7e6433d08c7101fa9e7e4b48/Firmware/LowLevel/src/datatypes.h#L177 ?
> But like to give it a more clear name then. Something like `WRONG_IJL` or however the current-sense-chip got called.

Using the bitmask sounds good. Maybe bits can be accessed more easily with an enum like you did with the halls? But for the name... your suggestion is very cryptic to me. Maybe `BROKEN_CHARGING_CURRENT_SENSOR`?

Reply:
I think I had a thought yesterday to have a "mask" to define which bytes/fields of the HL config shall be copied to LL. So instead of the memcpy, only copy the overridden bytes. Instead of saying what "undefined" means for each field and having at least one if-then-else per field.

Also not sure how flash would be handled, should it also store only the overridden fields?

But my mind is a bit blocked right now. If something else comes to mind, I'll share it with you (especially if it's easier), otherwise I'll see what you came up with. 😉

Reply:
> > Yes. In my vision also all other values like charge, battery, timeouts, ... if not separately defined in HL (mower_config).
> 
> But those wouldn't be changed "live"/at runtime by pressing some button on the mower, right?

No. This is only for volume (and in past also language change)... and I do not plan at the moment, that a volume change triggers a new config exchange. It only needs to be saved in flash so that a new power-on start with the right volume (and language).

The ones I'm mainly interested in, are only the default (or saved) ones and those get only exchanged when starting up or LowLevel got re-flashed.

> I think you're saying that you don't want to maintain the defaults in two different places, which I think is a good point. I wanted to make up my mind about this, and therefore wanted to know how "dynamic" the config is on LL side.
> 

LL has it's C500 "default" values from firmware, as well as the ones which are limited by hardware (limited by MOSFET or diode).
The "default" ones can be overwritten by HL for those people who use i.e. higher charging currents or different battery packs, not to forget the halls.
Those all get saved into flash (as you know), so that LL always boot up with some reasonable values.

I like to get it that way that HL doesn't need to know any of the LL values. LL shall inform HL about them.
And I also like to get it that way, that HL doesn't need to know about possible max ones. LL will change a possible to-large value on receive and send back in the return packet.

> > BTW: Think I know how to get rid of `IGNORE_CHARGING_CURRENT`. How's about giving it a bit in https://github.com/Apehaenger/OpenMower/blob/37f2ac6d2f30e52d7e6433d08c7101fa9e7e4b48/Firmware/LowLevel/src/datatypes.h#L177 ?
> > But like to give it a more clear name then. Something like `WRONG_IJL` or however the current-sense-chip got called.
> 
> Using the bitmask sounds good. Maybe bits can be accessed more easily with an enum like you did with the halls? But for the name... your suggestion is very cryptic to me. Maybe `BROKEN_CHARGING_CURRENT_SENSOR`?

Enum... hmm :thinking: , in the case of the halls it's really an "either or" so enum fits, but the bitmask is more a switch for a couple of things. Will take a look... probably a bitfield which would be more modern but should fit?!

Alternative `IGNORE_CHARGING_CURRENT` name:
Just checked, it's name is TSC101 :thinking: ... really not better :laughing: 
But yeees, much better! But it's not broken, simply wrong assembled. So `WRONG_CHARGING_CURRENT_SENSOR`? :+1:

Reply:
I think I understand your motivation. In short, we want to use the defaults and overlay them only with those values from HL which have been explicitly set. And the LL->HL config packet is mainly for displaying the actually used values.

Will take a step back and think about it.

> Enum... hmm 🤔

It was new to me, but you used [this](https://github.com/Apehaenger/OpenMower/blob/37f2ac6d2f30e52d7e6433d08c7101fa9e7e4b48/Firmware/LowLevel/src/datatypes.h#L161-L164). So I was wondering if you could have a list of `boolean xyz : 1` members? No idea whether that's valid syntax.

Reply:
> I think I understand your motivation. In short, we want to use the defaults and overlay them only with those values from HL which have been explicitly set. And the LL->HL config packet is mainly for displaying the actually used values.

Correctly :+1: .
We need them not only for minor stuff like sensor view, but also for major stuff like battery charge level which decide when to go back home.
Beside those, the LL->HL config packet is also to informing HL about probably to large values (requested before from HL) due to hardware limits.

> 
> Will take a step back and think about it.
> 
> > Enum... hmm 🤔
> 
> It was new to me, but you used [this](https://github.com/Apehaenger/OpenMower/blob/37f2ac6d2f30e52d7e6433d08c7101fa9e7e4b48/Firmware/LowLevel/src/datatypes.h#L161-L164). So I was wondering if you could have a list of `boolean xyz : 1` members? No idea whether that's valid syntax.

Yes :+1: . But I'm also unsure at the moment how the code would look. But as far as I know, bitfields got implemented exactly for such a case like ours. Insofar, probably good hint to push me away from the good old binary masking to some more modern stuff :laughing:

Reply:
My dominant thought is: We should have considered this when we talked all the time about how to define the struct, do the memcpy, automatically apply/keep defaults etc. while keeping compatibility between versions. The solution we finally agreed on feels elegant to me - but when HL is the source of truth, not when you just want to send a small subset of the fields. 😕 Especially the little magic with the memcpy isn't used at all.

Would it really be so bad to set defaults in the firmware that are safe for "all" mowers, and define either the same or more mower-specific values in HL? That would avoid uncertainties about priorities (is HL, LL, flash leading?), simplify the code (despite a bit more config), with the additional advantage that updated defaults could be included in a new image, which is much more likely to get applied than a new firmware. The part with the LL->HL communication still makes sense to me to see the actually used values after validation.

Reply:
I'm sorry, I though it's clear because it's how the config packet is live since months :innocent: 
Beside that, I asked to discuss how that existing packet can be made enlargeable without tackling with LL/HL compatibility (my old packet-trailer idea).
And to be clear, I'm still exited about the result of that discussion because we do now have a perfect flexible and compatible solution.

Don't worry about the packet/struct. It's pretty clear, straightforward, and since I changed now the packet- trigger- side from LL to HL (still local here), it became once more simpler (at least in my point of view :laughing: ).

But let's wait if I succeed, and then take a look. If it turns out that it's bad and to complicated, then it simply get trashed ;-)

Will give an intermediate commit once I do have two config values ready, so that you can take a view how it look